python/mpt/src/mpt/trie_diff.py

```diff
@@ -0,0 +1,388 @@
+                    post = None if post_int == 0 else U256(post_int)
+                    if address not in state_diff._storage_tries:
+                        state_diff._storage_tries[address] = {}
+                    state_diff._storage_tries[address][key] = tuple((pre, post))
+
+            state_diff._main_trie[address] = tuple((pre_account, post_account))
+
+        return state_diff
+
+    @classmethod
+    def from_tries(cls, tries: EthereumTrieTransitionDB) -> "StateDiff":
+        # Merge all mappings of hash -> node to have a single DB to fetch nodes and resolve addresses from
+        diff = StateDiff(
+            {}, {}, tries.nodes, tries.address_preimages, tries.storage_key_preimages
+        )
+
+        l_root = tries.state_root
+        r_root = tries.post_state_root
+
+        diff._compute_diff(l_root, r_root, Bytes(), diff._process_account_diff)
+        return diff
+
+    def _compute_diff(
+        self,
+        left: Optional[Hash32 | Bytes],
+        right: Optional[Hash32 | Bytes],
+        path: Bytes,
+        process_leaf_diff: Callable,
+    ):
+        if left == right:
+            return
+
+        def resolve(node: Hash32 | Bytes | None | InternalNode) -> InternalNode | None:
+            if isinstance(node, InternalNode):
+                return node
+            if not node or node == b"":
+                return None
+            if len(node) == 32:
+                return self._nodes.get(node)
+            return rlp.decode(node)
+
+        l_node = resolve(left)
+        r_node = resolve(right)
+
+        # Use direct class pattern matching
+        match (l_node, r_node):
+            case (None, None):
+                # No change
+                pass
+
+            case (None, LeafNode()):
+                # new leaf
+                preimage = nibble_path_to_bytes(path + r_node.rest_of_key)
+                process_leaf_diff(preimage, None, r_node)
+
+            case (None, ExtensionNode()):
+                # Look for diffs in the right sub-tree
+                self._compute_diff(
+                    None, r_node.subnode, path + r_node.key_segment, process_leaf_diff
+                )
+
+            case (None, BranchNode()):
+                # Look for diffs in all branches of the right sub-tree
+                for i in range(0, 16):
+                    self._compute_diff(
+                        None,
+                        r_node.subnodes[i],
+                        path + i.to_bytes(1, "big"),
+                        process_leaf_diff,
+                    )
+
+            case (LeafNode(), None):
+                # deleted leaf (should not happen post-cancun)
+                preimage = nibble_path_to_bytes(path + l_node.rest_of_key)
+                process_leaf_diff(preimage, l_node, None)
+
+            case (LeafNode(), LeafNode()):
+                if l_node.value != r_node.value:
+                    preimage = nibble_path_to_bytes(path + l_node.rest_of_key)
```

Comment:
should we check an invariant here, that l_node rest of key == r_node rest of key?